comms/src/control_service/worker.rs

```diff
@@ -180,31 +192,41 @@ where
             .map_err(|err| ControlServiceError::MessageError(err))?;
 
         let envelope_header: MessageEnvelopeHeader = envelope.to_header()?;
-        let message: Message = envelope.message_body()?;
+        if !envelope_header.flags.contains(MessageFlags::ENCRYPTED) {
+            return Err(ControlServiceError::ReceivedUnencryptedMessage);
+        }
+
+        let decrypted_body = envelope.body_frame(); // self.decrypt_body(envelope.body_frame(), &envelope_header.source)?;
+        let message =
+            Message::from_binary(decrypted_body.as_bytes()).map_err(ControlServiceError::MessageFormatError)?;
 
-        // TODO: Add outbound message service, peer etc.
         let context = ControlServiceMessageContext {
             envelope_header,
             message,
             connection_manager: self.connection_manager.clone(),
-            peer_manager: self.peer_manager.clone(),
+            peer_manager: self.connection_manager.get_peer_manager(),
         };
 
-        // TODO: Decryption of message
         debug!(target: LOG_TARGET, "Dispatching message");
         self.dispatcher.dispatch(context).map_err(|e| e.into())
     }
 
+    fn decrypt_body(&self, body: &Frame, public_key: &CommsPublicKey) -> Result<Frame> {
+        let ecdh_shared_secret = CommsPublicKey::shared_secret(&self.node_identity.secret_key, public_key).to_vec();
+        let ecdh_shared_secret_bytes: [u8; 32] = ByteArray::from_bytes(&ecdh_shared_secret)
+            .map_err(|_| ControlServiceError::SharedSecretSerializationError)?;
+        Ok(chacha20::decode(body, &ecdh_shared_secret_bytes))
```

Comment:
Instead of using `chacha20` directly, rather use `CommsCipher` from `types.rs`

Reply:
Ah see that's been merged - I'll rebase my code and update